core/src/main/kotlin/fr/sncf/osrd/standalone_sim/etcsContextBuilder.kt

```diff
@@ -67,11 +67,55 @@ fun buildETCSDangerPoints(
     return res.sorted()
 }
 
+/**
+ * Builds the list of ETCS detectors on the path. Goes until the last detector before the end of the
+ * path or, if the last block is ETCS, until the first ETCS detector after the end of the path.
+ */
+fun buildETCSBlockDetectors(
+    infra: FullInfra,
+    blockPath: StaticIdxList<Block>,
+    chunkPath: ChunkPath
+): List<Offset<TravelledPath>> {
+    val etcsBlockDetectors = mutableListOf<Offset<BlockPath>>()
+    var currentOffset = Offset.zero<BlockPath>()
+    if (isETCSBlock(blockPath.first(), infra)) etcsBlockDetectors.add(currentOffset)
+    for (block in blockPath) {
+        currentOffset += infra.blockInfra.getBlockLength(block).distance
+        if (isETCSBlock(block, infra)) etcsBlockDetectors.add(currentOffset)
+    }
+    val pathOffsetBuilder =
+        PathOffsetBuilder(
+            trainPathBlockOffset(infra.rawInfra, infra.blockInfra, blockPath, chunkPath).distance
+        )
+    if (etcsBlockDetectors.isEmpty()) return mutableListOf()
+    val etcsBlockDetectorsPathOffsets =
+        etcsBlockDetectors.map { pathOffsetBuilder.toTravelledPath(it) }
+    val res =
+        etcsBlockDetectorsPathOffsets
+            .filter { it.distance > Distance.ZERO && it.distance <= chunkPath.length }
+            .toMutableList()
+    // If the last block of the path is ETCS, return its last detector, even if it's after the end
+    // of the path.
+    if (
+        res.isNotEmpty() &&
+            etcsBlockDetectorsPathOffsets.last().distance >= chunkPath.length &&
+            res.last().distance < chunkPath.length
+    )
+        res.add(etcsBlockDetectorsPathOffsets.first { it.distance >= chunkPath.length })
```

Comment:
I don't get why filtering on the `chunkPath.length` (self-notice: which is equal to travelled-path length) is useful: the last block is part of the path, and we want its exit-detector if ETCS. Maybe I'm missing something, though.
I also think that it wouldn't be a problem to have multiple block-exits after the travelled-path, should it happen.

So it looks like we could simplify to:
```suggestion
    val res =
        etcsBlockDetectorsPathOffsets
            .filter { it.distance > Distance.ZERO }
            .toList()
```

Also, the current code will bug on a single-block path (full ETCS) as `res` would be empty when we want to add back the last detector, no?

Reply:
> I don't get why filtering on the chunkPath.length (self-notice: which is equal to travelled-path length) is useful: the last block is part of the path, and we want its exit-detector if ETCS. Maybe I'm missing something, though.
I also think that it wouldn't be a problem to have multiple block-exits after the travelled-path, should it happen.

I think there's a corner case where the path ends at the transition between the second to last block and the last block. In this case, we don't want the last block path detector. @eckter can you confirm?

> Also, the current code will bug on a single-block path (full ETCS) as res would be empty when we want to add back the last detector, no?

True. I should check `(res.isEmpty() && etcsBlockDetectorsPathOffsets.isNotEmpty()) || (etcsBlockDetectorsPathOffsets.last().distance >= chunkPath.length && res.last().distance < chunkPath.length)`.

See https://github.com/OpenRailAssociation/osrd/pull/12457/commits/7af9ad6e3d7d9fe63a4f903a612488d5ae47b3fe for now, might change according to future comments.

Reply:
Simplified the whole thing. We need the exit AND potentially entry detectors. So now i just return all of the etcs detectors as long as they are after the start of the path. See: https://github.com/OpenRailAssociation/osrd/pull/12457/commits/6fe5130924153b13a8503abce217e4dfdd333912